sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -567,6 +567,29 @@ public Duration getAllowedTimestampSkew() {
   @Target(ElementType.METHOD)
   public @interface ProcessElement {}
 
+  /**
+   * <b><i>Experimental - no backwards compatibility guarantees. The exact name or usage of this
+   * feature may change.</i></b>
+   *
+   * <p>Annotation that may be added to a {@link ProcessElement} or {@link OnTimer} method to
+   * indicate that the runner must ensure that the observable contents of the input {@link
+   * PCollection} or mutable state must be stable upon retries.
```

Comment:
input *and* mutable state must be stable?

Does stability of input require stability of bundling? What about if a timer fired after N elements, but on retry fires after N+K elements?

Reply:
Stability of input does no require stability of bundling. Indeed, the suggestion to allow this on `@OnTimer` was an interesting one since it brings up all kinds of issues surrounding state. Do we actually have to commit durably after every element? It would seem so, but that is unworkable. So given that stateful DoFn is very useful for sinks, and this annotation is also primarily for sinks, there is some real design discussion remaining.

Reply:
Brought this back to the dev list, since I don't think it will change the contents of this PR, and it will connect with ongoing discussions surrounding ordering of GBK iterables, etc.